src/policy/fees_util.cpp

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2024 The Bitcoin Core developers
+// Distributed under the MIT software license. See the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+
+#include <policy/fees_util.h>
+#include <policy/policy.h>
+
+BlockPercentiles CalculateBlockPercentiles(
+    const std::vector<std::tuple<CFeeRate, uint64_t>>& fee_rate_stats)
+{
+    if (fee_rate_stats.empty()) return BlockPercentiles();
+
+    auto start = fee_rate_stats.crbegin();
+    auto end = fee_rate_stats.crend();
+
+    unsigned int total_weight = 0;
+    const unsigned int p5_Size = static_cast<unsigned int>(0.05 * DEFAULT_BLOCK_MAX_WEIGHT);
+    const unsigned int p25_Size = static_cast<unsigned int>(0.25 * DEFAULT_BLOCK_MAX_WEIGHT);
+    const unsigned int p50_Size = static_cast<unsigned int>(0.5 * DEFAULT_BLOCK_MAX_WEIGHT);
+    const unsigned int p75_Size = static_cast<unsigned int>(0.75 * DEFAULT_BLOCK_MAX_WEIGHT);
```

Comment:
In  [policy, fees]: create a mempool based fee rate forecaster:

Perhaps we should call these `p*_weight`?